src/FormParserFactory.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace React\Http;
+
+use React\Http\Parser\FormUrlencoded;
+use React\Http\Parser\Multipart;
+use React\Http\Parser\NoBody;
+use React\Http\Parser\ParserInterface;
+
+class FormParserFactory
+{
+    /**
+     * @param Request $request
+     * @return ParserInterface
+     */
+    public static function create(Request $request)
+    {
+        $headers = $request->getHeaders();
+        $headers = array_change_key_case($headers, CASE_LOWER);
+
+        if (!array_key_exists('content-type', $headers)) {
+            return new NoBody($request);
+        }
+
+        $contentType = strtolower($headers['content-type']);
+
+        if (strpos($contentType, 'multipart/') === 0) {
+            return new Multipart($request);
+        }
+
+        if (strpos($contentType, 'application/x-www-form-urlencoded') === 0) {
+            return new FormUrlencoded($request);
+        }
+
+        return new NoBody($request);
```

Comment:
> The main concern I have is that there is currently no way to detect the en of the body if content length is missing. 

That's ok. If no content-length is given, ignore the body (use NoBody). But if given we should read the body until end of content-length and add it for example to a `RawBody` instance. I guess `NoBody` should only be used if we can not detect how long the body might be (either no Content-Length or no Content-Type).

No exception please if we can't detect a body because this is still valid http. I guess all of your current classes like NoBody make perfectly sense, however, we need a addition class that addresses the need of a content-types reactphp/http is not aware of or when the client sends a body (content-length > 0) but we can not decode it.

With that I can check for those 2 instances (NoBody, RawBody or else) to know whether I need to use the parser